File: adapters/python/src/project/scan_policy.py
from __future__ import annotations

from dataclasses import dataclass
from fnmatch import fnmatchcase


@dataclass(frozen=True)
class ScanPolicy:
    include: tuple[str, ...]
    exclude: tuple[str, ...]

    def allows(self, path: str) -> bool:
        for pattern in self.include:
            if fnmatchcase(path, pattern):
                return True

        for pattern in self.exclude:
            if fnmatchcase(path, pattern):
                return False

        return True

    def filter(self, paths: tuple[str, ...]) -> tuple[str, ...]:
        return tuple(path for path in paths if self.allows(path))

File: adapters/php/src/Project/ScanPolicy.php
<?php

declare(strict_types=1);

namespace Refactorlah\PhpAdapter\Project;

use function preg_match;
use function preg_quote;
use function array_filter;
use function array_values;
use function str_replace;

final class ScanPolicy
{
    /**
     * @param list<string> $include
     * @param list<string> $exclude
     */
    public function __construct(
        private readonly array $include,
        private readonly array $exclude,
    ) {}

    public function allows(string $path): bool
    {
        foreach ($this->include as $pattern) {
            if ($this->matches($pattern, $path)) {
                return true;
            }
        }

        foreach ($this->exclude as $pattern) {
            if ($this->matches($pattern, $path)) {
                return false;
            }
        }

        return true;
    }

    /**
     * @param list<string> $files
     * @return list<string>
     */
    public function filter(array $files): array
    {
        return array_values(array_filter($files, fn(string $file): bool => $this->allows($file)));
    }

    private function matches(string $pattern, string $path): bool
    {
        $quoted = preg_quote($pattern, '/');
        $quoted = str_replace('\*\*', '.*', $quoted);
        $quoted = str_replace('\*', '[^/]*', $quoted);
        $quoted = str_replace('\?', '[^/]', $quoted);

        return 1 === preg_match('/^' . $quoted . '$/', $path);
    }
}
